src/OpenTelemetry/Logs/OpenTelemetryLoggerProvider.cs

```diff
@@ -45,48 +46,49 @@ public OpenTelemetryLoggerProvider(IOptionsMonitor<OpenTelemetryLoggerOptions> o
 
         internal OpenTelemetryLoggerProvider(OpenTelemetryLoggerOptions options)
         {
-            this.options = options ?? throw new ArgumentNullException(nameof(options));
-            this.loggers = new Dictionary<string, ILogger>(StringComparer.Ordinal);
+            this.Options = options ?? throw new ArgumentNullException(nameof(options));
 
             foreach (var processor in options.Processors)
             {
                 this.AddProcessor(processor);
             }
         }
 
-        internal IExternalScopeProvider ScopeProvider
-        {
-            get
-            {
-                if (this.scopeProvider == null)
-                {
-                    this.scopeProvider = new LoggerExternalScopeProvider();
-                }
-
-                return this.scopeProvider;
-            }
-        }
-
         void ISupportExternalScope.SetScopeProvider(IExternalScopeProvider scopeProvider)
         {
-            // TODO: set existing loggers
             this.scopeProvider = scopeProvider;
+
+            foreach (KeyValuePair<string, OpenTelemetryLogger> logger in this.loggers)
+            {
+                logger.Value.ScopeProvider = scopeProvider;
+            }
         }
 
         public ILogger CreateLogger(string categoryName)
         {
-            lock (this.loggers)
+            while (true)
             {
-                ILogger logger;
+                if (this.loggers.TryGetValue(categoryName, out var logger))
+                {
+                    return logger;
+                }
+
+                logger = new OpenTelemetryLogger(categoryName, this)
+                {
+                    ScopeProvider = this.scopeProvider,
+                };
 
-                if (this.loggers.TryGetValue(categoryName, out logger))
+                if (this.loggers.TryAdd(categoryName, logger))
```

Comment:
Do we know if `this.loggers.TryGetValue(categoryName, out var logger)` will always succeed when `TryAdd` failed?
If yes, the code can be simplified by returning `this.loggers[categoryName]`.
If no, perhaps store the logger locally to avoid creating multiple instances in the loop?

Reply:
@reyang I updated it (again) to use `Hashtable`. @mconnew just [pointed out to me](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/38#discussion_r580611766) that `Hashtable` has unique [thread safety gurantees](https://docs.microsoft.com/en-us/dotnet/api/system.collections.hashtable?view=net-5.0#thread-safety) which make it useful for caching scenarios.

Reply:
Looks great! 💯